File: backend/src/routes/adminRoutes.js
// routes/adminRoutes.js
import express from 'express';
import {
    createUser, listUsers, deleteUser, updateUser, getUserById,
    createGroup, listGroups, deleteGroup, updateGroup,
    createRole, listRoles, deleteRole, updateRole,
    createMenu, listMenus, deleteMenu, updateMenu, assignMenusToRole, updateMenusOfRole,
    removeUserFromGroup,
    removeRoleFromGroup,
    removeMenuFromRole,
    getUserGroups, assignGroupsToUser, getGroupRoles,
    assignRolesToGroup, getRoleMenus,
    getUsersWithGroups, getRolesWithGroups, getRolesOfGroupWithGroupId,
    updateRolesOfGroup

} from '../controllers/adminController.js';
import { login } from '../middlewares/authenticate.js';
import { upload } from '../middlewares/upload.js';

const router = express.Router();

// User routes
router.post('/users', upload.single('avatar'), createUser);
router.get('/users', listUsers);
router.delete('/users/:id', deleteUser);
router.put("/users/:id", upload.single("avatar"), updateUser);
router.get('/users/:id', getUserById);
router.post('/login', login);

// Group routes
router.post('/groups', createGroup);
router.get('/groups', listGroups);
router.delete('/groups/:id', deleteGroup);
router.put('/groups/:id', updateGroup);

// Role routes
router.post('/roles', createRole);
router.get('/roles', listRoles);
router.delete('/roles/:id', deleteRole);
router.put('/roles/:id', updateRole);

// Menu routes
router.post('/menus', createMenu);
router.get('/menus', listMenus);
router.delete('/menus/:id', deleteMenu);
router.put('/menus/:id', updateMenu);

// Assignments user to group
router.post('/user-group/:id', assignGroupsToUser);
router.get('/user-groups', getUsersWithGroups);
router.delete('/user-group/remove', removeUserFromGroup);

// Assignments group to role
router.post('/group-roles/:id', assignRolesToGroup);
router.get('/group-roles', getRolesWithGroups);
router.get('/groupid-role/:id', getRolesOfGroupWithGroupId);
router.delete('/group-role/remove', removeRoleFromGroup);
router.post('/groupid-roles/:id', updateRolesOfGroup); 

// Assignments role to menu

router.post('/role-menu/:id', assignMenusToRole);
router.get('/role-menus', getRoleMenus);
router.delete('/role-menu/:role_id', removeMenuFromRole);
router.post('/role-menus/:id', updateMenusOfRole);



export default router;

File: backend/app/routes/admin_routes.py
from flask import Blueprint
from controllers import admin_controller
from middlewares.authenticate import login, require_api_key
from middlewares.upload import upload_file

admin_bp = Blueprint('admin_bp', __name__)

# User routes
admin_bp.route('/users', methods=['POST'])(require_api_key(upload_file(admin_controller.create_user)))
admin_bp.route('/users', methods=['GET'])(require_api_key(admin_controller.list_users))
admin_bp.route('/users/<int:user_id>', methods=['DELETE'])(require_api_key(admin_controller.delete_user))
admin_bp.route('/users/<int:user_id>', methods=['PUT'])(require_api_key(upload_file(admin_controller.update_user)))
admin_bp.route('/users/<int:user_id>', methods=['GET'])(require_api_key(admin_controller.get_user_by_id))
admin_bp.route('/login', methods=['POST'])(require_api_key(login))

# Group routes
admin_bp.route('/groups', methods=['POST'])(require_api_key(admin_controller.create_group))
admin_bp.route('/groups', methods=['GET'])(require_api_key(admin_controller.list_groups))
admin_bp.route('/groups/<int:group_id>', methods=['DELETE'])(require_api_key(admin_controller.delete_group))
admin_bp.route('/groups/<int:id>', methods=['PUT'])(require_api_key(admin_controller.update_group))

# Role routes
admin_bp.route('/roles', methods=['POST'])(require_api_key(admin_controller.create_role))
admin_bp.route('/roles', methods=['GET'])(require_api_key(admin_controller.list_roles))
admin_bp.route('/roles/<int:role_id>', methods=['DELETE'])(require_api_key(admin_controller.delete_role))
admin_bp.route('/roles/<int:role_id>', methods=['PUT'])(require_api_key(admin_controller.update_role))

# Menu routes
admin_bp.route('/menus', methods=['POST'])(require_api_key(admin_controller.create_menu))
admin_bp.route('/menus', methods=['GET'])(require_api_key(admin_controller.list_menus))
admin_bp.route('/menus/<int:id>', methods=['DELETE'])(require_api_key(admin_controller.delete_menu))
admin_bp.route('/menus/<int:id>', methods=['PUT'])(require_api_key(admin_controller.update_menu))

# Assign user to group
admin_bp.route('/user-group/<int:user_id>', methods=['POST'])(require_api_key(admin_controller.assign_groups_to_user))
admin_bp.route('/user-groups', methods=['GET'])(require_api_key(admin_controller.get_users_with_groups))
admin_bp.route('/user-group/remove', methods=['DELETE'])(require_api_key(admin_controller.remove_user_from_group))

# Assign role to group
admin_bp.route('/group-roles/<int:group_id>', methods=['POST'])(require_api_key(admin_controller.assign_roles_to_group))
admin_bp.route('/group-roles', methods=['GET'])(require_api_key(admin_controller.get_roles_with_groups))
admin_bp.route('/groupid-role/<int:group_id>', methods=['GET'])(require_api_key(admin_controller.get_roles_of_group_with_group_id))
admin_bp.route('/group-role/remove', methods=['DELETE'])(require_api_key(admin_controller.remove_role_from_group))
admin_bp.route('/groupid-roles/<int:group_id>', methods=['POST'])(require_api_key(admin_controller.update_roles_of_group))

# Assign menu to role
admin_bp.route('/role-menu/<int:role_id>', methods=['POST'])(require_api_key(admin_controller.assign_menus_to_role))
admin_bp.route('/role-menus', methods=['GET'])(require_api_key(admin_controller.get_role_menus))
admin_bp.route('/role-menu/<int:role_id>', methods=['DELETE'])(require_api_key(admin_controller.remove_menu_from_role))
admin_bp.route('/role-menus/<int:role_id>', methods=['POST'])(require_api_key(admin_controller.update_menus_of_role))
